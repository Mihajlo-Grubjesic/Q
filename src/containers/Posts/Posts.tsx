import { useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Card } from '../../components/Card/Card';
import {
  CardsWrapper,
  ContentWrapper,
} from '../../components/Card/styled/Card';
import { withMessage } from '../../components/HOC/withMessage';
import { Input } from '../../components/Input/Input';
import { NavigationLink } from '../../components/Link/Link';
import { ROUTES } from '../../constants/routes';
import { useDebounce } from '../../hooks/useDebounce';
import { PostData } from '../../types';
import { NoPostsResults, PostsSearch } from './components/PostsList/PostsList';

interface Props {
  posts: PostData[];
  message?: string;
}

const Posts = ({ posts, message }: Props): JSX.Element => {
  console.log(`${message} Posts List`);
  const { t } = useTranslation();

  const [filteredPosts, setFilteredPosts] = useState(posts);
  const [searchTerm, setSearchTerm] = useState('');

  const handlePostSearch = (searchTerm: string) => {
    setSearchTerm(searchTerm);
  };

  const debouncedSearchTerm = useDebounce(searchTerm, 500);

  useEffect(() => {
    if (debouncedSearchTerm === '') {
      setFilteredPosts(posts);
    } else {
      const filterPosts = posts.filter((post) =>
        post.username.toLowerCase().includes(debouncedSearchTerm.toLowerCase()),
      );
      setFilteredPosts(filterPosts);
    }
  }, [debouncedSearchTerm, posts]);

  return (
    <>
      <PostsSearch>
        <Input
          onChange={(e) => handlePostSearch(e.target.value)}
          placeholder={t('searchPostsPlaceholder') as string}
        />
      </PostsSearch>
      {filteredPosts.length === 0 && (
        <NoPostsResults>No results</NoPostsResults>
      )}
      <CardsWrapper>
        {filteredPosts.map(({ id, title, username, comments }) => (
          <NavigationLink to={`${ROUTES.POSTS}/${id}`} key={id}>
            <Card
              title={title}
              subtitle={username}
              content={
                <ContentWrapper>
                  {comments.map((comment, index) => (
                    <p key={index}>‚óè {comment}</p>
                  ))}
                </ContentWrapper>
              }
            />
          </NavigationLink>
        ))}
      </CardsWrapper>
    </>
  );
};

export default withMessage(Posts);
