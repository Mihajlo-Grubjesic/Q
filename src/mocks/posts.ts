import { Post } from '../types';

export const PostsMock: Post[] = [
  {
    id: 1,
    userId: 1,
    body: 'Post 1',
    title: 'Post 1',
  },
  {
    id: 2,
    userId: 2,
    body: 'Post 2',
    title: 'Post 2',
  },
  {
    id: 3,
    userId: 3,
    body: 'Post 3',
    title: 'Post 3',
  },
  {
    id: 4,
    userId: 4,
    body: 'Post 4',
    title: 'Post 4',
  },
];
