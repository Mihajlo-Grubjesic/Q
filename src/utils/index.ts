import { Post, User, Comment, PostData } from "../types";

export const combinePostsData = ({
  posts,
  users,
  comments,
}: {
  posts: Post[];
  users: User[];
  comments: Comment[];
}): PostData[] => {
  return posts.map((post) => {
    const username =
      users.find((user) => user.id === post.userId)?.username ?? "";
    const postComments = comments
      .filter((comment) => comment.postId === post.id)
      .map((comment) => comment.body);
    return { ...post, username, comments: postComments };
  });
};
