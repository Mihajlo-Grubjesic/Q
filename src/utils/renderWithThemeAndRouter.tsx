import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { DefaultTheme, ThemeProvider } from 'styled-components';
import { lightTheme } from '../theme/themes';

export function renderWithThemeAndRouter(
  children: React.ReactElement,
  theme?: DefaultTheme,
) {
  return render(
    <ThemeProvider theme={theme ? theme : lightTheme}>
      <Router>{children}</Router>
    </ThemeProvider>,
  );
}
