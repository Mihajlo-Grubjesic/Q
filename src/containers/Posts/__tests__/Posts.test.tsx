import { vi } from 'vitest';
import { fireEvent, screen, waitFor } from '@testing-library/react';
import { renderWithThemeAndRouter } from '../../../utils/renderWithThemeAndRouter';

import { CommentsMock } from '../../../mocks/comments';
import { PostsMock } from '../../../mocks/posts';
import { UsersMock } from '../../../mocks/users';
import { combinePostsData } from '../../../utils';
import Posts from '../Posts';

vi.mock('react-i18next', () => ({
  useTranslation: () => ({ t: (key: string) => key }),
}));
describe('Posts', () => {
  const posts = combinePostsData({
    posts: PostsMock,
    users: UsersMock,
    comments: CommentsMock,
  });
  it('renders posts and test click on post redirection', () => {
    const { container } = renderWithThemeAndRouter(<Posts posts={posts} />);

    const clickedPostIdMock = 3;
    const clickedPostMock = posts.find((post) => post.id === clickedPostIdMock);
    fireEvent.click(screen.getByText(clickedPostMock?.title ?? ''));
    expect(window.location.pathname).toEqual(`/posts/${clickedPostMock?.id}`);

    expect(container).toMatchSnapshot();
  });

  it('test posts no results filter action', async () => {
    renderWithThemeAndRouter(<Posts posts={posts} />);
    const searchPostInput = screen.getByRole('textbox');

    fireEvent.change(searchPostInput, {
      target: { value: 'No existing username' },
    });
    await waitFor(() => {
      expect(screen.getByText('No results')).toBeInTheDocument();
    });
  });
});
