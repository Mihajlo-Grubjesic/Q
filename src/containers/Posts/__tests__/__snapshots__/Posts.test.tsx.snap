// Vitest Snapshot v1

exports[`Posts > renders posts and test click on post redirection 1`] = `
<div>
  <div
    class="sc-iBYQkv cMOQto"
  >
    <input
      class="sc-jSUZER dHbPHT"
      placeholder="searchPostsPlaceholder"
      value=""
    />
  </div>
  <div
    class="sc-bcXHqe ilmuKn"
  >
    <a
      class="sc-gKPRtg eaKZcD"
      href="/posts/1"
    >
      <div
        class="sc-dkrFOg ysqzP"
      >
        <div
          class="sc-hLBbgP kaMPmA"
        >
          Post 1
        </div>
        <div
          class="sc-eDvSVe cwgtvh"
        >
          Bret
        </div>
        <div
          class="sc-gswNZR huITSX"
        >
          <p>
            ● 
            laudantium enim quasi est quidem magnam voluptate ipsam eos
tempora quo necessitatibus
dolor quam autem quasi
reiciendis et nam sapiente accusantium
          </p>
          <p>
            ● 
            est natus enim nihil est dolore omnis voluptatem numquam
et omnis occaecati quod ullam at
voluptatem error expedita pariatur
nihil sint nostrum voluptatem reiciendis et
          </p>
          <p>
            ● 
            quia molestiae reprehenderit quasi aspernatur
aut expedita occaecati aliquam eveniet laudantium
omnis quibusdam delectus saepe quia accusamus maiores nam est
cum et ducimus et vero voluptates excepturi deleniti ratione
          </p>
          <p>
            ● 
            non et atque
occaecati deserunt quas accusantium unde odit nobis qui voluptatem
quia voluptas consequuntur itaque dolor
et qui rerum deleniti ut occaecati
          </p>
          <p>
            ● 
            harum non quasi et ratione
tempore iure ex voluptates in ratione
harum architecto fugit inventore cupiditate
voluptates magni quo et
          </p>
        </div>
      </div>
    </a>
    <a
      class="sc-gKPRtg eaKZcD"
      href="/posts/2"
    >
      <div
        class="sc-dkrFOg ysqzP"
      >
        <div
          class="sc-hLBbgP kaMPmA"
        >
          Post 2
        </div>
        <div
          class="sc-eDvSVe cwgtvh"
        >
          Antonette
        </div>
        <div
          class="sc-gswNZR huITSX"
        >
          <p>
            ● 
            doloribus at sed quis culpa deserunt consectetur qui praesentium
accusamus fugiat dicta
voluptatem rerum ut voluptate autem
voluptatem repellendus aspernatur dolorem in
          </p>
          <p>
            ● 
            maiores sed dolores similique labore et inventore et
quasi temporibus esse sunt id et
eos voluptatem aliquam
aliquid ratione corporis molestiae mollitia quia et magnam dolor
          </p>
          <p>
            ● 
            ut voluptatem corrupti velit
ad voluptatem maiores
et nisi velit vero accusamus maiores
voluptates quia aliquid ullam eaque
          </p>
          <p>
            ● 
            sapiente assumenda molestiae atque
adipisci laborum distinctio aperiam et ab ut omnis
et occaecati aspernatur odit sit rem expedita
quas enim ipsam minus
          </p>
          <p>
            ● 
            voluptate iusto quis nobis reprehenderit ipsum amet nulla
quia quas dolores velit et non
aut quia necessitatibus
nostrum quaerat nulla et accusamus nisi facilis
          </p>
        </div>
      </div>
    </a>
    <a
      class="sc-gKPRtg eaKZcD"
      href="/posts/3"
    >
      <div
        class="sc-dkrFOg ysqzP"
      >
        <div
          class="sc-hLBbgP kaMPmA"
        >
          Post 3
        </div>
        <div
          class="sc-eDvSVe cwgtvh"
        >
          Samantha
        </div>
        <div
          class="sc-gswNZR huITSX"
        >
          <p>
            ● 
            ut dolorum nostrum id quia aut est
fuga est inventore vel eligendi explicabo quis consectetur
aut occaecati repellat id natus quo est
ut blanditiis quia ut vel ut maiores ea
          </p>
          <p>
            ● 
            expedita maiores dignissimos facilis
ipsum est rem est fugit velit sequi
eum odio dolores dolor totam
occaecati ratione eius rem velit
          </p>
          <p>
            ● 
            fuga eos qui dolor rerum
inventore corporis exercitationem
corporis cupiditate et deserunt recusandae est sed quis culpa
eum maiores corporis et
          </p>
          <p>
            ● 
            vel quae voluptas qui exercitationem
voluptatibus unde sed
minima et qui ipsam aspernatur
expedita magnam laudantium et et quaerat ut qui dolorum
          </p>
          <p>
            ● 
            nihil ut voluptates blanditiis autem odio dicta rerum
quisquam saepe et est
sunt quasi nemo laudantium deserunt
molestias tempora quo quia
          </p>
        </div>
      </div>
    </a>
    <a
      class="sc-gKPRtg eaKZcD"
      href="/posts/4"
    >
      <div
        class="sc-dkrFOg ysqzP"
      >
        <div
          class="sc-hLBbgP kaMPmA"
        >
          Post 4
        </div>
        <div
          class="sc-eDvSVe cwgtvh"
        >
          Karianne
        </div>
        <div
          class="sc-gswNZR huITSX"
        >
          <p>
            ● 
            iste ut laborum aliquid velit facere itaque
quo ut soluta dicta voluptate
error tempore aut et
sequi reiciendis dignissimos expedita consequuntur libero sed fugiat facilis
          </p>
          <p>
            ● 
            consequatur necessitatibus totam sed sit dolorum
recusandae quae odio excepturi voluptatum harum voluptas
quisquam sit ad eveniet delectus
doloribus odio qui non labore
          </p>
          <p>
            ● 
            veritatis voluptates necessitatibus maiores corrupti
neque et exercitationem amet sit et
ullam velit sit magnam laborum
magni ut molestias
          </p>
          <p>
            ● 
            doloribus est illo sed minima aperiam
ut dignissimos accusantium tempore atque et aut molestiae
magni ut accusamus voluptatem quos ut voluptates
quisquam porro sed architecto ut
          </p>
          <p>
            ● 
            qui harum consequatur fugiat
et eligendi perferendis at molestiae commodi ducimus
doloremque asperiores numquam qui
ut sit dignissimos reprehenderit tempore
          </p>
        </div>
      </div>
    </a>
  </div>
</div>
`;
